### PR TITLE
Add MaxResponseDrainSize, and fix a few draining bugs

### PR DESCRIPTION
**Pull request by stephentoub**
I've added MaxResponseDrainSize to make the setting configurable rather than a const; I left it internal until the API is approved.

In adding tests for it, I found a few existing bugs in draining that I fixed:
- HttpConnectionResponseContent nulls out its Stream when the Stream is consumed.  Then when the response content is disposed, the Stream is null, so it's not disposed, so it's not drained.
- HttpContentReadStream's Dispose wasn't checking whether it was already called, and would kick off a new draining operation if the connection was still associated.  So if the stream was disposed multiple times, which happens if the response content and stream are both disposed (since the content disposes of the stream), we end up corrupting the response due to multiple threads concurrently accessing and draining it.

cc: @geoffkizer, @davidsh, @wfurt
Contributes to https://github.com/dotnet/corefx/issues/27329

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs [hunk 1]
```diff
@@ -9,94 +9,106 @@
...
+                // Only attempt draining if we haven't started draining due to disposal; otherwise
+                // multiple calls to Dispose (which happens frequently when someone disposes of the
+                // response stream and response content) will kick off multiple concurrent draining
+                // operations. Also don't delegate to the base if Dispose has already been called,
+                // as doing so will end up disposing of the connection before we're done draining.
+                if (_disposed)
+                {
+                    return;
+                }
+                _disposed = true;
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs]: Should the check/set be atomic for thread safety?

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs]: Yeah, probably.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ResponseDrain.cs [hunk 2]
```diff
@@ -162,7 +160,15 @@ private static string GetResponseForContentMode(string content, ContentMode mode
                     string response = GetResponseForContentMode(content, mode);
                     await server.AcceptConnectionAsync(async connection =>
                     {
-                        await connection.ReadRequestHeaderAndSendCustomResponseAsync(response);
+                        // Process the first request, with some introduced delays in the response to
+                        // stress the draining.
+                        await connection.ReadRequestHeaderAsync().ConfigureAwait(false);
+                        foreach (char c in response)
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ResponseDrain.cs]: Could use DribbleStream here.  Maybe not worth it.

Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ResponseDrain.cs]: I started down that path, but it didn't have existing Read/ReadAsync/CopyToAsync implementations, and decided this was just easier.

Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ResponseDrain.cs]: Ok

**Comment by geoffkizer** [comment]
LGTM, minor feedback above

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Outerloop Linux x64 Debug Build please

**Comment by stephentoub** [comment]
@dotnet-bot test NETFX x86 Release Build please
@dotnet-bot test OSX x64 Debug Build please

**Pull request merged**